docs/build/guides/dapps/state-archival.mdx

```diff
@@ -0,0 +1,348 @@
+    const fee = StellarSdk.BASE_FEE;
+
+    const transaction = new StellarSdk.TransactionBuilder(account, {
+      fee,
+      networkPassphrase,
+    })
+      .addOperation(contract.call(method, ...args))
+      .setTimeout(30)
+      .build();
+
+    let preparedTransaction = await server.prepareTransaction(transaction);
+
+    let signedXDR = await signTransaction(preparedTransaction.toXDR());
+    let signedTransaction = StellarSdk.TransactionBuilder.fromXDR(
+      signedXDR,
+      networkPassphrase,
+    );
+
+    // Try to send the transaction
+    const sim = await server.simulateTransaction(signedTransaction);
+
+    // Other failures are out of scope of this tutorial.
+    if (!Api.isSimulationSuccess(sim)) {
+      throw sim;
+    }
+
+    // If simulation didn't fail, we don't need to restore anything! Just send it.
+    if (Api.isSimulationRestore(sim)) {
+      console.log("Data archived. Attempting restoration...");
+
+      // Prepare restoration transaction
+      const restoreTx = new StellarSdk.TransactionBuilder(account, { fee })
+        .setNetworkPassphrase(networkPassphrase)
+        .addOperation(StellarSdk.Operation.restoreFootprint({}))
+        .setTimeout(30)
+        .build();
+
+      let preparedRestoreTx = await server.prepareTransaction(restoreTx);
+      let signedRestoreXDR = await signTransaction(preparedRestoreTx.toXDR());
+      let signedRestoreTx = StellarSdk.TransactionBuilder.fromXDR(
+        signedRestoreXDR,
+        networkPassphrase,
+      );
+
+      await server.sendTransaction(signedRestoreTx);
+      console.log("Restoration complete. Retrying original transaction...");
+
+      // Retry the original transaction
+      return submitOrRestoreAndRetry(contractId, method, ...args);
+    }
+
+    const result = await server.sendTransaction(signedTransaction);
+
+    return result;
+  } catch (error) {
+    console.error("Transaction failed:", error);
+    throw error;
+  }
+}
+```
+
+This function now uses Freighter for signing transactions. It first checks if Freighter is connected and authorized, then proceeds with the transaction. If restoration is needed (indicated by a `HostStorageError`), it creates a separate restoration transaction, signs it with Freighter, and submits it before retrying the original transaction.
+
+#### Step 3: Use the helper function in your dapp
+
+You can now use this function to make contract calls that automatically handle restoration:
+
+```javascript
+async function performContractAction(contractId, method, ...args) {
+  try {
+    const result = await submitOrRestoreAndRetry(contractId, method, ...args);
+    console.log("Transaction successful:", result);
+    return result;
+  } catch (error) {
+    console.error("Error performing contract action:", error);
+    // Handle the error appropriately in your UI
+  }
+}
+```
+
+#### Step 4: Handling contract instance restoration
+
+For restoring an entire contract instance, you might need a separate function:
```

Comment:
It'd be good to cross-reference the [existing example](https://developers.stellar.org/docs/data/rpc/api-reference/methods/getLedgerEntries#requesting-a-contracts-wasm-code) here so users can learn more!